### ISSUE TITLE
ParlAI Core H1 2020 Roadmap

### ISSUE THREAD

**Issue by stephenroller** [facebookresearch/ParlAI#2228]
# ParlAI H1 2020 Roadmap

## Community preface

This is an experiment in sharing our roadmap with the OSS community. The hope is that users can be aware of major changes that might be upcoming, or suggest modifications that they might find valuable. Future roadmaps may or may not be shared. The community is welcome to participate in the development of this roadmap, and in its completion. If OSS users wish to work on some of the tasks, they should say so on the issue first to avoid conflicts.

This document contains a rough sketch of our planned efforts for the first half of 2020. It is focused entirely on the engineering aspects of ParlAI, and does not contain any of our research plans. The roadmap is a work in progress, and issues may be added to it later. We revisit our roadmaps at roughly 6-8 week intervals to check on progress, and revise expectations. We do not expect to fully complete every item on the roadmap.

This document only represents the roadmap for things roughly including parlai/core and modeling agents. We have separate roadmap documents for the ParlAI Messenger and ParlAI MTurk sides of this repo. Their roadmaps will not be shared at this time.

## Intro

Our mission remains similar to past halves: we want to be the best platform available for doing dialogue research, to facilitate our own research, and to share our research and tools with the OSS community.

Our broad goals for the half:
- Speedups and improvements in training of models, especially to improve scaling.
- Improved software engineering practices to facilitate better engineering.
- Improved usability via major improvements to the ubiquitous `Opt`

## Summary of issues

P0...P3 indicate priorities, with P0 being mission critical and P3 being lowest priority.

S0...S3 indicate scope of the change, with S0 being a major revision to behavior, and S3 being a minor isolated change.

### Complete

Issue | Status | Title | Priority | Scope | Type
-------|-------|-------|----------|------|----
#2238 | Complete | Speed up CI testing                                                              | P1 | Large    |                     
#2181 | Complete | New LR Schedulers                                                                | P1 | Medium   | enhancement         
#2164 | Complete | Incremental decoding in TransformerDecoder                                       | P1 | Medium   | enhancement         
#2219 | Complete | Move safety classifier test to nightly                                           | P3 | Minor    | tests               
#2162 | Complete | Metrics revisiting                                                               | P0 | Grand    | enhancement         
#2192 | Complete | Dynamic Batching                                                                 | P1 | Grand    | enhancement         
#1876 | Complete | ParlAI auto-generated docs script for all standard agents                        | P1 |          | docs                
#2131 | Complete | [TorchAgent] FP16 tweaks                                                         | P2 | Minor    | enhancement         
#2169 | Complete | Consolidate `.valid`, `.best_valid` and `.trainstats` files                      | P2 | Minor    | enhancement         
#2300 | Complete | Improved PR annotations for flake8, mypy, etc                                    | P2 | Medium   |                     
#2106 | Complete | [tensorboard] Resuming model logging causes bad resume in tensorboard            | P1 |          | bug                 
#1988 | Complete | GPT2 Agent                                                                       | P1 |   Large       | Feature             
#2236 | Complete | interactive.py testing                                                           | P1 | Medium   |                     
#1977 | Complete | [TA] report gnorm even when no clipping                                          | P1 | Minor    | enhancement         
#2227 | Complete | Tweak issue templates                                                            | P3 | Minor    | codequality         
#2121 | Complete | [codecov] not reporting all lines                                                | P2 |          |                     
#2197 | Complete | Unit test for StopTrain                                                          | P2 | Minor    |                     
#2087 | Complete | --underscores_in_options  is not behaving correctly                              | P2 | Minor    |                     
#2168 | Complete | Throw error if bool is used in add_argument?                                     | P2 | Minor    |                     
#2063 | Complete | Improve interactive_web                                                          | P2 |          | help wanted         
#2172 | Complete | Opt save/load                                                                    | P1 | Minor    | enhancement         
#2163 | Complete | Type checking                                                                    | P2 | Grand    | enhancement         
#2175 | Complete | Set up stale PR actions                                                          | P3 | Minor    | codequality         
#2185 | Complete | Easier multi-gpu usage                                                           | P1 | Medium    | enhancement         
#2188 | Complete | ParlAI `jsonl` teacher                                                           | P0 | Large    | enhancement         
#2166 | Complete | Global parlai command line launcher                                              | P2 | Medium   |                     

### In Progress

Issue | Status | Title | Priority | Scope | Type
-------|-------|-------|----------|------|----
#1918 | Partial | Add additional tests for Torch Ranker Agents                                     | P2 |          | help wanted         
#2215 | Partial | HuggingFace Transformers interop                                                 | P2 | Grand    | enhancement         

### Planned

Issue | Status | Title | Priority | Scope | Type
-------|-------|-------|----------|------|----
#2165 | Planned | Immutable opt                                                                    | P0 | Grand    | enhancement         
#2325 | Planned | Teacher Mocking                                               | P1 | Grand | codequality
#2221 | Planned | Self-Chat Improvements                                                           | P2 |          | enhancement         
#2177 | Planned | ParlaiParser.add_argument should support no_save and no_override                 | P2 | Large    | enhancement         
#2176 | Planned | Opt should remember where options were defined                                   | P2 | Medium   | enhancement         
#2299 | Planned | More transformer variants                                                        | P2 | Large    |                     
#2187 | Planned | [BatchWorld] Eliminate batch padding episodes                                    | P3 | Medium   | codequality         
#2170 | Planned | Unify names of some options                                                      | P3 | Minor    | codequality         
#2186 | Planned | [TA] Add batchsize to the Batch object                                           | P3 | Minor    | enhancement         

# Abandoned

Issue | Status | Title | Priority | Scope | Type
-------|-------|-------|----------|------|----
#2184 | Abandoned | Support tied positional weights in transformers                                  | P1 | Medium   | enhancement         
#2103 | Abandoned | [transformer] Switch to torch.gelu                                               | P1 | Minor    | codequality         
#2141 | Abandoned | Outputs of train_model/eval_model should be json objects.                        | P2 | Minor    | enhancement         
#2222 | Abandoned | Support for pytorch quantization                                                 | P3 | Grand    | enhancement         
#2115 | Abandoned | Remove classmethod awkward super()'s.                                            | P2 | Minor    | codequality         
#2179 | Abandoned | Investigate whether match_batch is still needed                                  | P3 | Medium   | codequality         
#2191 | Abandoned | Preprocessing support for Agents                                                 | P1 | Grand    | enhancement

**Comment by stephenroller** [comment]
Enumerated all the current issues, there will likely be more additions in the near term.
